sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,502 @@
+              connectionConfiguration.getKeyspace(),
+              table);
+
+      ArrayList<TokenRange> tokenRanges = new ArrayList<>();
+      for (Row row : resultSet) {
+        TokenRange tokenRange =
+            new TokenRange(
+                row.getLong("partitions_count"),
+                row.getLong("mean_partition_size"),
+                row.getString("range_start"),
+                row.getString("range_end"));
+        tokenRanges.add(tokenRange);
+      }
+      // The table may not contain the estimates yet
+      // or have partitions_count and mean_partition_size fields = 0
+      // if the data was just inserted and the amount of data in the table was small.
+      // This is very common situation during tests,
+      // when we insert a few rows and immediately query them.
+      // However, for tiny data sets the lack of size estimates is not a problem at all,
+      // because we don't want to split tiny data anyways.
+      // Therefore, we're not issuing a warning if the result set was empty
+      // or mean_partition_size and partitions_count = 0.
+      return tokenRanges;
+    } catch (InvalidQueryException e) {
+      LOG.warn(
+          "Failed to fetch size estimates for $keyspaceName.$tableName "
+              + "from system.size_estimates table. "
+              + "The number of created bundles may be inaccurate. "
+              + "Please make sure you use Cassandra 2.1.5 or newer.");
+      ArrayList<TokenRange> tokenRanges = new ArrayList<>();
+      tokenRanges.add(new TokenRange(0L, 0L, "0", "0"));
+      return tokenRanges;
+    }
+  }
+
+  /**
+   * Measure distance between two tokens.
+   *
+   * @param cluster The cassandra {@link Cluster} object
+   * @param token1 The measure is symmetrical so token1 and token2 can be exchanged
+   * @param token2 The measure is symmetrical so token1 and token2 can be exchanged
+   * @return Number of tokens that separate token1 and token2
+   */
+  private static BigInteger distance(Cluster cluster, BigInteger token1, BigInteger token2) {
+    Partitioner partitioner = new Partitioner(cluster);
+    //token2 > token1
+    if (token2.compareTo(token1) == 1) {
+      return token2.subtract(token1);
+    } else {
+      return token2.subtract(token1).add(partitioner.getTotalTokens());
+    }
+  }
+
+  static class Partitioner {
+    BigInteger minToken;
+    BigInteger maxToken;
+
+    Partitioner(Cluster cluster) {
+      String partitionerName = cluster.getMetadata().getPartitioner();
+      switch (partitionerName) {
+        case "org.apache.cassandra.dht.Murmur3Partitioner":
+          minToken = new BigInteger(String.valueOf(Long.MIN_VALUE));
+          maxToken = new BigInteger(String.valueOf(Long.MAX_VALUE));
+          break;
+        case "org.apache.cassandra.dht.RandomPartitioner":
+          minToken = new BigInteger("-1");
+          maxToken = new BigInteger(String.valueOf((long) Math.pow(2, 127)));
```

Comment:
I'm pretty sure that converting Math.pow(2, 127) into a long is lossy, so you likely end up having maxToken not actually represent the maximum token, and having data loss.

Also... why not just use Long.MIN_VALUE and MAX_VALUE in all cases? Then perhaps you can get rid of this whole class, and of the BigInteger machinery (instead use long)